[#113] Fix for application/x-www-form-urlencoded

afronski opened the pull request:
Main point of this fix is to introduce proper payload documentation for OpenAPI 2.0 (changing type of parameter for a given Content-Type from query to formData) and OpenAPI 3.0 (moving payload definition from query parameters to requestBody).

It does not fix schema generation for OpenAPI 3.0 (it is still generating empty schema, although - it generates proper example if provided).

If merged, this closes #113

ozscheyge commented on restdocs-api-spec-openapi-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi2/OpenApi20GeneratorTest.kt:
```diff
@@ -230,6 +230,26 @@ class OpenApi20GeneratorTest {
     }
 
+    @Test
+    fun `should handle urlencoded body in POST request as a formData not query string`() {
```
...in POST request as formData

ozscheyge commented on restdocs-api-spec-openapi-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi2/OpenApi20GeneratorTest.kt:
```diff
@@ -832,6 +878,41 @@ class OpenApi20GeneratorTest {
+                        type = OAUTH2,
+                        requiredScopes = listOf("prod:c")
+                ),
+                headers = listOf(
```
This should not be needed

ozscheyge commented on restdocs-api-spec-openapi-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi2/OpenApi20GeneratorTest.kt:
```diff
@@ -832,6 +878,41 @@ class OpenApi20GeneratorTest {
+                method = method,
+                contentType = "application/x-www-form-urlencoded",
+                schema = schema,
+                securityRequirements = SecurityRequirements(
```
This should not be needed

ozscheyge commented on restdocs-api-spec-openapi-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi2/OpenApi20GeneratorTest.kt:
```diff
@@ -230,6 +230,26 @@ class OpenApi20GeneratorTest {
+    }
+
+    @Test
+    fun `should handle urlencoded body in PUT request as a formData not query string`() {
```
...in PUT request as formData

ozscheyge commented on restdocs-api-spec-openapi-generator/src/main/kotlin/com/epages/restdocs/apispec/openapi2/OpenApi20Generator.kt:
```diff
@@ -412,7 +420,7 @@ object OpenApi20Generator {
     }
 
-    private fun requestParameterDescriptor2Parameter(parameterDescriptor: ParameterDescriptor): QueryParameter {
+    private fun requestParameterDescriptor2QueryParameter(parameterDescriptor: ParameterDescriptor): QueryParameter {
```
This could be combined with the function below by using generics, but the current solution is also fine and expressive

afronski replied:
Great, so I will leave it as is.

ozscheyge commented on restdocs-api-spec-openapi3-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi3/OpenApi3GeneratorTest.kt:
```diff
@@ -730,6 +778,30 @@ class OpenApi3GeneratorTest {
         )
     }
 
+    private fun getProductRequestAsFormData(method: HTTPMethod, getSecurityRequirement: () -> SecurityRequirements = ::getOAuth2SecurityRequirement): RequestModel {
```
The `get` in `getProductRequestAsFormData` comes from the HTTP method GET. Thus I'd just name it
`productRequestAsFormData` (the method is defined by the first parameter)

ozscheyge commented on restdocs-api-spec-openapi3-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi3/OpenApi3GeneratorTest.kt:
```diff
@@ -730,6 +778,30 @@ class OpenApi3GeneratorTest {
 
+    private fun getProductRequestAsFormData(method: HTTPMethod, getSecurityRequirement: () -> SecurityRequirements = ::getOAuth2SecurityRequirement): RequestModel {
+        return RequestModel(
+                path = "/products/{id}-{subId}",
```
This is probably due to copy&paste from another given method, but does just distract here. I'd simplify it to:
`path = "/products/{id}",`

ozscheyge commented on restdocs-api-spec-openapi3-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi3/OpenApi3GeneratorTest.kt:
```diff
@@ -178,14 +178,47 @@ class OpenApi3GeneratorTest {
     }
 
+    @Test
+    fun `should handle urlencoded body in POST request as a request body not query string`() {
```
should treat urlencoded body in POST request as request body and not as query

ozscheyge commented on restdocs-api-spec-openapi3-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi3/OpenApi3GeneratorTest.kt:
```diff
@@ -178,14 +178,47 @@ class OpenApi3GeneratorTest {
+    }
+
+    @Test
+    fun `should handle urlencoded body in PUT request as a request body not query string`() {
```
should treat urlencoded body in PUT request as request body and not as query

ozscheyge commented on restdocs-api-spec-openapi3-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi3/OpenApi3GeneratorTest.kt:
```diff
@@ -178,14 +178,47 @@ class OpenApi3GeneratorTest {
+    }
+
+    fun thenResourceHasFormDataInRequestBodyNotAsQueryParameters(method: String) {
+        val productGetByIdPath = "paths./products/{id}-{subId}.$method"
```
This would need to be adapted to `/products/{id}` if the suggestion below is followed

ozscheyge commented on restdocs-api-spec-openapi3-generator/src/test/kotlin/com/epages/restdocs/apispec/openapi3/OpenApi3GeneratorTest.kt:
```diff
@@ -178,14 +178,47 @@ class OpenApi3GeneratorTest {
+        thenResourceHasFormDataInRequestBodyNotAsQueryParameters(method.toString().toLowerCase())
+    }
+
+    fun thenResourceHasFormDataInRequestBodyNotAsQueryParameters(method: String) {
```
thenResourceHasFormDataInRequestBody**And**NotAsQueryParameters

afronski commented:
@ozscheyge I have introduced improvements requested by you, and also added the implementation for the aforementioned case with empty schema response for OpenAPI 3.0. 

Feel free to review once again and provide feedback. If implementation is okay, I think it should definitely close #113 without a need for introducing additional issues after.

bigDDDeveloper commented:
Hello @afronski : Kindly confirm the jar version in which this issue is fixed as i am getting empty object schema with "com.epages:restdocs-api-spec-mockmvc:0.11.4"
